### DISCUSSION TITLE
classification of the relationship between two sentences

### DISCUSSION THREAD

**Discussion by SimonF89** [explosion/spaCy#5354] [New Features & Project Ideas]
Hi, everybody,

I would like to automatically classify the relation between two requirements. Here I would like to recognize if:
- Both requirements have nothing to do with each other.
   - The house must be yellow --> The lawn must be green.
- Both requirements are similar or equal.
   - The house must be yellow --> The colour of the house must be yellow.
- Both requirements are contradictory.
  - The house must be yellow. --> The color of the house must be green.

I first tried the implementation of Sentence-Transformers and used this data set here https://nlp.stanford.edu/blog/the-stanford-nli-corpus-revisited/ for training. Unfortunately the result was not as hoped for.

I would now like to generate training data myself. Do you have a comparable function in Spacy, which I can use for the classification of two sets of requirements? Cause if i would like to get valid results using the sentence-transformer solution, i would need tausends of samples...

**Comment by adrianeboyd** [comment]
No, spacy doesn't really have any support for models that work with pairs of texts at this point.

**Comment by SimonF89** [comment]
any recommadations what i could use to train the contradictory problem? Similarty works quite fine with sentence-transformers.

**Comment by adrianeboyd** [comment]
No, sorry!

**Comment by SimonF89** [comment]
ok, thx anyway!

**Comment by SimonF89** [comment]
i think i found something:
https://github.com/free-variation/spaCy/blob/master/examples/notebooks/Decompositional%20Attention.ipynb

**Comment by svlandeg** [comment]
Good to hear you've found some interesting pointers, and sorry we can't be of more help!

In fact, you might have more succes on StackOverflow with these kind of generic NLP usage questions - a bigger community reads those :-)

I'll close this here for now as there's not really an action point for us to follow up on.

**Reply by stevemarin** [reply]
Are these responses still valid?  I'm looking to attempt 0-shot relation extraction using some MLMs tuned on MNLI (or some other natural language inference dataset).  Could concatenating the premise and hypothesis with the appropriate token (< / s >, [SEP], or whatever) and running through a pre-trained transformer model be a reasonable approach?  Is there a way to keep/use the classification head on a pre-trained transformer model that I'm not seeing?  Or would I need to append my own classifier to the transformer?

**Reply by polm** [reply]
We haven't done any work on this since then, no. 

You could try concatenating things; I'm not sure it wouldn't work but I wouldn't be too optimistic.

> Is there a way to keep/use the classification head on a pre-trained transformer model that I'm not seeing? Or would I need to append my own classifier to the transformer?

The spacy-transformers approach only uses Transformers for features, classifications heads are not used. You can implement a custom component to use them if you want to.
